[PT FE] Add aten::as_strided

mmikolajcz opened the pull request:
### Details:
 - *Add aten::as_strided*
 - *...*

### Tickets:
 - *101107*

andrei-kochin commented:
@mvafin 
could you please take a look?

mvafin commented on src/frontends/pytorch/src/op/as_strided.cpp:
```diff
@@ -0,0 +1,82 @@
+    if (std::dynamic_pointer_cast<v0::Constant>(context.get_input_from_visible_context(1).get_node_shared_ptr())) {
+        auto input_vector = context.const_input<std::vector<int64_t>>(1);
+        std::for_each(input_vector.rbegin(), input_vector.rend(), [&](int64_t input_val) {
+            auto const_input = v0::Constant::create(element::i32, Shape{}, {input_val});
```
Could you please call `mark_node` for each `Constant`.

mmikolajcz replied:
Done

mvafin reviewed (changes requested):
This code produce accuracy problems on models with `aten::as_strided`.
Consider this layer test:
```
class TestAsStridedLongformer(PytorchLayerTest):
    def _prepare_input(self):
        return (np.random.randn(1, 10, 20, 40).astype(np.float32).transpose([0,2,3,1]),)

    def create_model(self):
        class aten_as_strided_lf(torch.nn.Module):
            def forward(self, x):
                chunk_size = list(x.size())
                chunk_size[1] = chunk_size[1] * 2 - 1
                chunk_stride = list(x.stride())
                chunk_stride[1] = chunk_stride[1] // 2
                return x.as_strided(size=chunk_size, stride=chunk_stride)

        ref_net = None

        return aten_as_strided_lf(), ref_net, "aten::as_strided"

    @pytest.mark.nightly
    @pytest.mark.precommit
    def test_as_strided_lf(self, ie_device, precision, ir_version):
        self._test(*self.create_model(), ie_device, precision, ir_version, trace_model=True, freeze_model=True)
```
It fails by accuracy

mmikolajcz commented:
> This code produce accuracy problems on models with `aten::as_strided`. Consider this layer test:
> 
> ```
> class TestAsStridedLongformer(PytorchLayerTest):
>     def _prepare_input(self):
>         return (np.random.randn(1, 10, 20, 40).astype(np.float32).transpose([0,2,3,1]),)
> 
>     def create_model(self):
>         class aten_as_strided_lf(torch.nn.Module):
>             def forward(self, x):
>                 chunk_size = list(x.size())
>                 chunk_size[1] = chunk_size[1] * 2 - 1
>                 chunk_stride = list(x.stride())
>                 chunk_stride[1] = chunk_stride[1] // 2
>                 return x.as_strided(size=chunk_size, stride=chunk_stride)
> 
>         ref_net = None
> 
>         return aten_as_strided_lf(), ref_net, "aten::as_strided"
> 
>     @pytest.mark.nightly
>     @pytest.mark.precommit
>     def test_as_strided_lf(self, ie_device, precision, ir_version):
>         self._test(*self.create_model(), ie_device, precision, ir_version, trace_model=True, freeze_model=True)
> ```
> 
> It fails by accuracy

Fixed by retrieving strides information from torchscript, this should now work but it will require for model to be traced and not frozen. Added this example as test case

mvafin commented:
Could you also add models with `as_strided` to precommit testing:

- Add here https://github.com/openvinotoolkit/openvino/blob/a00d28aac0efda832209dbaf6794481a1c16005c/tests/model_hub_tests/torch_tests/test_hf_transformers.py#L295
- Remove xfail here to add to nightly here https://github.com/openvinotoolkit/openvino/blob/a00d28aac0efda832209dbaf6794481a1c16005c/tests/model_hub_tests/torch_tests/hf_transformers_models#L13 and here https://github.com/openvinotoolkit/openvino/blob/a00d28aac0efda832209dbaf6794481a1c16005c/tests/model_hub_tests/torch_tests/hf_transformers_models#L304

mmikolajcz commented:
> Could you also add models with `as_strided` to precommit testing:
> 
> * Add here https://github.com/openvinotoolkit/openvino/blob/a00d28aac0efda832209dbaf6794481a1c16005c/tests/model_hub_tests/torch_tests/test_hf_transformers.py#L295
> * Remove xfail here to add to nightly here https://github.com/openvinotoolkit/openvino/blob/a00d28aac0efda832209dbaf6794481a1c16005c/tests/model_hub_tests/torch_tests/hf_transformers_models#L13
>    and here https://github.com/openvinotoolkit/openvino/blob/a00d28aac0efda832209dbaf6794481a1c16005c/tests/model_hub_tests/torch_tests/hf_transformers_models#L304

Done, removed xfails and added "allenai/led-base-16384" to precommit since it seem to be quicker to load/infer than those from nightly xfailed models

mvafin commented on src/frontends/pytorch/src/op/as_strided.cpp:
```diff
@@ -0,0 +1,106 @@
+    std::for_each(strides.rbegin(), strides.rend(), [&](Output<Node>& stride) {
+        auto const_num_iter = context.mark_node(v0::Constant::create(element::i32, Shape{1}, {strides_size - i}));
+        stride = context.mark_node(std::make_shared<v0::Convert>(stride, element::i32));
+        auto size = context.mark_node(std::make_shared<v0::Convert>(sizes.at(strides_size - i), element::i32));
```
No need to convert to same type to use Range
```suggestion
        auto size = sizes.at(strides_size - i);
```
